File: HelloSwiftTour/Test/Tools/WCSecurityTool.m
//
//  WCSecurityTool.m
//  Test
//
//  Created by wesley_chen on 2022/4/28.
//

#import "WCSecurityTool.h"
#import <CommonCrypto/CommonCryptor.h>

@implementation WCSecurityTool

+ (nullable NSData *)AES256CryptWithOperation:(CCOperation)operation data:(NSData *)data key:(NSString *)key  {
    if (![data isKindOfClass:[NSData class]] || ![key isKindOfClass:[NSString class]]) {
        return nil;
    }
    
    // 'key' should be 32 bytes for AES256, will be null-padded otherwise
    char keyPtr[kCCKeySizeAES256+1]; // room for terminator (unused)
    bzero(keyPtr, sizeof(keyPtr)); // fill with zeroes (for padding)
    
    // fetch key data
    [key getCString:keyPtr maxLength:sizeof(keyPtr) encoding:NSUTF8StringEncoding];
    
    NSUInteger dataLength = [data length];
    
    //See the doc: For block ciphers, the output size will always be less than or
    //equal to the input size plus the size of one block.
    //That's why we need to add the size of one block here
    size_t bufferSize = dataLength + kCCBlockSizeAES128;
    void *buffer = malloc(bufferSize);
    
    size_t numBytes = 0;
    CCCryptorStatus cryptStatus = CCCrypt(operation, kCCAlgorithmAES, kCCOptionECBMode | kCCOptionPKCS7Padding,
                                          keyPtr, kCCKeySizeAES256,
                                          NULL /* initialization vector (optional) */,
                                          [data bytes], dataLength, /* input */
                                          buffer, bufferSize, /* output */
                                          &numBytes);
    if (cryptStatus == kCCSuccess) {
        //the returned NSData takes ownership of the buffer and will free it on deallocation
        return [NSData dataWithBytesNoCopy:buffer length:numBytes];
    }
    
    free(buffer); //free the buffer;
    return nil;
}

+ (nullable NSData *)AES256EncryptWithData:(NSData *)data key:(NSString *)key {
    return [self AES256CryptWithOperation:kCCEncrypt data:data key:key];
}

+ (nullable NSData *)AES256DecryptWithData:(NSData *)data key:(NSString *)key {
    return [self AES256CryptWithOperation:kCCDecrypt data:data key:key];
}

@end

File: HelloSecurity/WCSecurityTool/WCSecurityTool.swift
//
//  WCSecurityTool.swift
//  HelloSecurity
//
//  Created by wesley_chen on 2022/4/26.
//

import UIKit
import CommonCrypto
// https://stackoverflow.com/a/67498866
class WCSecurityTool {
    class func encryptionAESModeECB(messageData data: Data, key: String) -> Data? {
        guard let keyData = key.data(using: String.Encoding.utf8) else { return nil }
        guard let cryptData = NSMutableData(length: Int((data.count)) + kCCBlockSizeAES128) else { return nil }
        
        let keyLength               = size_t(kCCKeySizeAES256)
        let keyDataM = NSMutableData.init(length: keyLength)
        keyDataM?.setData(keyData)
        keyDataM?.length = keyLength // fill zero
        
        let operation:  CCOperation = UInt32(kCCEncrypt)
        let algoritm:   CCAlgorithm = UInt32(kCCAlgorithmAES)
        let options:    CCOptions   = UInt32(kCCOptionECBMode + kCCOptionPKCS7Padding)
        let iv:         String      = ""
        
        var numBytesEncrypted: size_t = 0
        
        let cryptStatus = CCCrypt(operation,
                                  algoritm,
                                  options,
                                  (keyDataM?.copy() as! NSData).bytes, keyLength,
                                  //(keyData as NSData).bytes, keyLength,
                                  iv,
                                  (data as NSData).bytes, data.count,
                                  cryptData.mutableBytes, cryptData.length,
                                  &numBytesEncrypted)
        
        if UInt32(cryptStatus) == UInt32(kCCSuccess) {
            cryptData.length = Int(numBytesEncrypted)
            return cryptData.copy() as? Data
        } else {
            return nil
        }
    }
    
    class func encryptionAESModeECB_2(messageData data: Data, key: String) -> Data? {
        guard let keyData = key.data(using: String.Encoding.utf8) else { return nil }
        guard let cryptData = NSMutableData(length: Int((data.count)) + kCCBlockSizeAES128) else { return nil }
        
        let keyLength               = size_t(kCCKeySizeAES128)
        let operation:  CCOperation = UInt32(kCCEncrypt)
        let algoritm:   CCAlgorithm = UInt32(kCCAlgorithmAES)
        let options:    CCOptions   = UInt32(kCCOptionECBMode + kCCOptionPKCS7Padding)
        let iv:         String      = ""
        
        var numBytesEncrypted: size_t = 0
        
        let cryptStatus = CCCrypt(operation,
                                  algoritm,
                                  options,
                                  (keyData as NSData).bytes, keyLength,
                                  iv,
                                  (data as NSData).bytes, data.count,
                                  cryptData.mutableBytes, cryptData.length,
                                  &numBytesEncrypted)
        
        if UInt32(cryptStatus) == UInt32(kCCSuccess) {
            cryptData.length = Int(numBytesEncrypted)
            let encryptedString = cryptData.base64EncodedString(options: .lineLength64Characters)
            return encryptedString.data(using: .utf8)
        } else {
            return nil
        }
    }
}
